### PR TITLE
Add property-based tests for tar and rsync builders (#71)

### LINKED ISSUE #71: Property-based tests for tar_create / rsync_sync command construction (cuprum/builders/tar.py)
## Overview

Add Hypothesis property-based tests for `tar_create` and `rsync_sync` in `cuprum/builders/tar.py` (around line 51).

## Justification

Command construction has flag-combination and ordering risk currently only covered by examples.

## Benefit

- Proves exactly-one compression flag
- Proves required argv shape
- Proves deterministic path conversion

## Cost Trade-Off

Low. Pure command vectors.

## Verifiability Improvements

Extract immutable argv-builder helpers before wrapping in `sh.make`.

## Recommended Order

Phase 2 — validators and builders.

---
Raised from PR https://github.com/leynos/cuprum/pull/62 by @leynos.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Closes #71

Adds Hypothesis property-based tests for the tar and rsync command builders:

- **Required argv shape**: `tar_create` produces exactly `-c`, an optional compression flag, `-f`, the archive, then the sources in order; `tar_extract` produces `-x -f <archive>` with `-C <destination>` only when supplied.
- **Exactly-one compression flag**: the flag matches the `Compression` member (`-z`/`-j`/`-J`) and never appears more than once; `Compression.NONE` emits none.
- **Fixed rsync flag order**: exactly the flags enabled on `RsyncOptions` appear, in the documented order, followed by source then destination.
- **Deterministic path conversion**: equal inputs give equal argv, and `Path` inputs agree with their string equivalents.
- **Rejection contracts**: empty source sequences raise `ValueError`; a bare `str`/`Path` where a sequence is required raises `TypeError`; relative paths require `allow_relative=True`.
- **Wrapper fidelity**: the public builders attach the curated `TAR`/`RSYNC` programs and the argv produced by the pure builders.

Per the issue's verifiability recommendation, the pure argv builders `_tar_create_argv`, `_tar_extract_argv`, and `_rsync_argv` are extracted (with an explicit `_FLAG_ORDER` table making the rsync flag sequence part of the stated contract); the public builders wrap them in `sh.make` and their behaviour is unchanged.

## Stacking

Based on `issue-119-tar-compression-enum` (PR #142, the `Compression` enum) atop `fix-ci-lint-ruff-version-skew` (PR #167); will rebase as those merge.

## Testing

- `make check-fmt`, `make lint`, `make typecheck` all pass
- `make test`: 571 passed, 43 skipped (Python); Rust suite passing
- CodeRabbit review: 0 findings

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Introduce a Compression enum for tar builders, refactor tar and rsync command construction into pure argv helpers, and add property-based tests to pin down their argv contracts.

New Features:
- Expose a Compression enum and updated TarCreateOptions API for selecting tar compression algorithms via a single choice.

Enhancements:
- Refactor tar_create, tar_extract, and rsync_sync to delegate to internal pure argv builder functions with a defined flag order.
- Adjust Makefile tasks and tooling integration to run ruff via uv and align local/CI tooling usage.

Build:
- Update Makefile tooling targets so ruff is treated as a venv tool and executed through uv-run commands.

CI:
- Stop installing ruff as a global uv tool in the CI workflow, relying instead on the venv-based setup.

Documentation:
- Document the Compression enum design and update user guide examples to use the new TarCreateOptions.compression API.

Tests:
- Add Hypothesis-based property tests for tar and rsync builders to verify argv shape, flag ordering, path handling, and wrapper behaviour.
- Adjust existing unit and behaviour tests to use the new Compression-based tar options instead of boolean flags.

## References

- Lody session: https://lody.ai/leynos/sessions/704b357f-024d-46b7-8a51-021f58ec1c01

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Excess Number of Function Arguments
cuprum/unittests/test_builder_property_based.py: test_path_conversion_is_deterministic_and_type_insensitive

### What lead to degradation?
test_path_conversion_is_deterministic_and_type_insensitive has 6 arguments, max arguments = 4

### Why does this problem occur?
Functions with many arguments indicate either a) low cohesion where the function has too many responsibilities, or b) a missing abstraction that encapsulates those arguments.
The threshold for the Python language is 4 function arguments.

### How to fix it?
Start by investigating the responsibilities of the function. Make sure it doesn't do too many things, in which case it should be split into smaller and more cohesive functions. Consider the refactoring [INTRODUCE PARAMETER OBJECT](https://refactoring.com/catalog/introduceParameterObject.html) to encapsulate arguments that refer to the same logical concept.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat documentation and validation coverage as in scope).

### ❌ Failed checks (1 error, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The property tests pin argv shape and order, but they never drive relative-path handling through `tar_create` or `rsync_sync`, so a skipped `safe_path` integration could still pass. | Add property cases that call `tar_create` and `rsync_sync` with relative inputs under `allow_relative=False/True`, so path-validation regressions fail. |
| Developer Documentation | ⚠️ Warning | docs/cuprum-design.md records Compression, but docs/developers-guide.md has no tar/rsync helper guidance, and the 3.1.1 execplan still describes the old gzip/bzip2/xz API. | Add developer-guide coverage for the new builder helpers and refresh or retire the stale 3.1.1 execplan so it matches the Compression API and argv-helper extraction. |

File: cuprum/unittests/test_builder_property_based.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    source: str,
+    destination: str,
+    options: RsyncOptions,
+) -> None:
+    """Exactly the enabled flags appear, in _FLAG_ORDER, then the paths."""
+    argv = _rsync_argv(source, destination, options)
+    expected_flags = tuple(flag for attr, flag in _FLAG_ORDER if getattr(options, attr))
+    assert argv == (*expected_flags, source, destination), (
+        "argv must be the enabled flags in fixed order, then source and destination"
+    )
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_builder_property_based.py:158]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: cuprum/unittests/test_builder_property_based.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    options: RsyncOptions,
+) -> None:
+    """Equal inputs give equal argv; Path and str inputs agree."""
+    rsync_from_strings = _rsync_argv(source, destination, options)
+    assert rsync_from_strings == _rsync_argv(
+        Path(source),
+        Path(destination),
+        options,
+    ), "rsync argv must agree between Path and str inputs"
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_builder_property_based.py:203]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by pandalump** [comment]
@coderabbitai resume

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Inline comments:
In `@cuprum/builders/rsync.py`:
- Around line 61-68: Update the public docstring for the rsync command builder
to use Oxford -ize spelling: replace “synchronisation” with “synchronization”
and “synchronise” with “synchronize,” while leaving the surrounding parameter
documentation unchanged.

In `@cuprum/builders/tar.py`:
- Around line 45-53: Update the public TarCreateOptions docstring with a
NumPy-style Attributes section documenting compression and allow_relative,
including each field’s type, purpose, and default value. Preserve the existing
summary and clarify that allow_relative applies to both archive and source
paths.
- Around line 77-82: Update the input validation in tar_create to classify bare
string or Path sources before checking emptiness, ensuring an empty bare path
raises the documented TypeError. Prefer structural pattern matching over
isinstance or imperative decomposition, then retain the existing ValueError for
genuinely empty source sequences.

In `@docs/cuprum-design.md`:
- Around line 261-262: Rewrite the sentence near the SafeCmd/DynamicCmd guidance
to remove the second-person pronoun, using third-person or imperative wording
while preserving the existing meaning.

In `@docs/developers-guide.md`:
- Around line 1013-1016: Insert a comma before “so” in the
compatibility-argument documentation sentence describing mixed usage raising
ValueError, leaving the surrounding wording unchanged.

---

Duplicate comments:
In `@docs/developers-guide.md`:
- Around line 20-25: Update the documentation sentence describing
`_tar_create_argv`, `_tar_extract_argv`, and `_rsync_argv` by inserting a comma
before “so” to correctly separate the independent clauses; leave the surrounding
wording unchanged.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat documentation and validation coverage as in scope).

### ❌ Failed checks (1 error, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The suite misses the bare-empty-string source case the review flagged, so `tar_create(archive, "")` would still slip through untested. | Add a separate test that passes an empty string as `sources` and asserts `TypeError`; keep the empty-sequence `ValueError` check distinct. |
| Domain Architecture | ⚠️ Warning | `_tar_create_argv` checks emptiness before type, so a bare empty string hits ValueError instead of the documented TypeError. | Move the bare `str`/`Path` shape check before the empty-sequence check, and add a test for `sources=""`. |
